File: Selenium/Activities/Activity21.java
package activities;

import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;
import java.util.Set;

public class Activity21 {
    public static void main(String[] args) throws InterruptedException {
        System.setProperty(FirefoxDriver.SystemProperty.BROWSER_LOGFILE, "/dev/null");
        WebDriverManager.firefoxdriver().setup();
        WebDriver driver = new FirefoxDriver();
        WebDriverWait wait=new WebDriverWait(driver, Duration.ofSeconds(10));
        driver.get("https://v1.training-support.net/selenium/tab-opener");
        System.out.println("Title of homepage:" + driver.getTitle());
        System.out.println("Parent tab: " + driver.getWindowHandle());
        driver.findElement(By.id("launcher")).click();

        wait.until(ExpectedConditions.numberOfWindowsToBe(2));
        System.out.println("Currently open windows: " + driver.getWindowHandles());
        for(String handle:driver.getWindowHandles()){
            driver.switchTo().window(handle);
        }
        wait.until(ExpectedConditions.titleIs("Newtab"));
        wait.until(ExpectedConditions.visibilityOfElementLocated(By.id("actionButton"))).click();
        System.out.println("Current tab: " + driver.getWindowHandle());
        System.out.println("Page title: " + driver.getTitle());
        System.out.println("Heading of web page:"+driver.findElement(By.xpath("//div[@class='content']")).getText());
        wait.until(ExpectedConditions.numberOfWindowsToBe(3));
        for(String handle : driver.getWindowHandles()) {
            driver.switchTo().window(handle);
        }
        wait.until(ExpectedConditions.titleIs("Newtab2"));
        System.out.println("Current tab: " + driver.getWindowHandle());
        System.out.println("Page title: " + driver.getTitle());
        System.out.println("Heading of web page:"+driver.findElement(By.xpath("//div[@class='content']")).getText());

        driver.quit();

    }
}

File: Selenium/Activity21.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.firefox.service import Service as FirefoxService
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.wait import WebDriverWait
from webdriver_manager.firefox import GeckoDriverManager

# Set up the Firefox Driver with WebDriverManger
service = FirefoxService(GeckoDriverManager().install())

# Start the Driver
with webdriver.Firefox(service=service) as driver:
    driver.get("https://v1.training-support.net/selenium/tab-opener")
    wait = WebDriverWait(driver, 10)
    print("Title of homepage:" + driver.title)
    print("Parent tab: " + str(driver.current_window_handle))
    driver.find_element(By.ID, "launcher").click()
    wait.until(expected_conditions.number_of_windows_to_be(2))
    print("Currently open windows: " + str(driver.window_handles))
    for handle in driver.window_handles:
        driver.switch_to.window(handle)
    wait.until(expected_conditions.title_is("Newtab"))
    wait.until(expected_conditions.visibility_of_element_located((By.ID, "actionButton"))).click()
    print("Current tab: " + driver.current_window_handle)
    print("Page title: " + driver.title)
    print("Heading of web page:" + driver.find_element(By.XPATH, "//div[@class='content']").text)
    wait.until(expected_conditions.number_of_windows_to_be(3))
    for handle in driver.window_handles:
        driver.switch_to.window(handle)
    wait.until(expected_conditions.title_is("Newtab2"))
    print("Current tab: " + driver.current_window_handle)
    print("Page title: " + driver.title)
    print("Heading of web page:" + driver.find_element(By.XPATH, "//div[@class='content']").text)

    driver.quit()
